src/openforms/formio/components/vanilla.py

```diff
@@ -563,24 +573,112 @@ def rewrite_for_request(component: ContentComponent, request: Request):
+        "empty": _("This list may not be empty."),
+        "min_length": _("Ensure this field has at least {min_length} elements."),
+        "max_length": _("Ensure this field has no more than {max_length} elements."),
+    }
+
+    def __init__(self, **kwargs):
+        self.registry = kwargs.pop("registry")
+        self.components: list[Component] = kwargs.pop("components", [])
+        self.allow_empty = kwargs.pop("allow_empty", True)
+        self.max_length = kwargs.pop("max_length", None)
+        self.min_length = kwargs.pop("min_length", None)
+        super().__init__(**kwargs)
+        if self.max_length is not None:
+            message = lazy_format(
+                self.error_messages["max_length"], max_length=self.max_length
+            )
+            self.validators.append(MaxLengthValidator(self.max_length, message=message))
+        if self.min_length is not None:
+            message = lazy_format(
+                self.error_messages["min_length"], min_length=self.min_length
+            )
+            self.validators.append(MinLengthValidator(self.min_length, message=message))
+
+    def get_value(self, dictionary):
+        if self.field_name not in dictionary:
+            if getattr(self.root, "partial", False):
+                return serializers.empty
+        # We don't bother with html
+        return dictionary.get(self.field_name, serializers.empty)
+
+    def to_internal_value(self, data):
+        """
+        List of dicts of native values <- List of dicts of primitive datatypes.
+        """
+        if isinstance(data, (str, Mapping)) or not hasattr(data, "__iter__"):
+            self.fail("not_a_list", input_type=type(data).__name__)
+        if not self.allow_empty and len(data) == 0:
+            self.fail("empty")
+        return self.run_child_validation(data)
+
+    def to_representation(self, value: list):
+        """
+        List of object instances -> List of dicts of primitive datatypes.
+        """
+        child = self._build_child()
+        return [
+            child.to_representation(item) if item is not None else None
+            for item in value
+        ]
+
+    def _build_child(self, **kwargs):
+        return build_serializer(
+            components=copy.deepcopy(self.components),
             # XXX: check out type annotations here, there's some co/contra variance
             # in play
             register=self.registry,
+            **kwargs,
         )
+
+    def run_child_validation(self, data):
+        result = []
+        errors = {}
+
+        for idx, item in enumerate(data):
+            # given the local scope of data, build a nested serializer for the component
+            # configuration and apply the dynamic hide/visible logic to it.
+            # Note that we add the editgrid key as container so that the
+            # conditional.when values resolve, as these look like `editgridparent.child`.
```

Comment:
this is similar to what formio does, however they do it the other way around - the find the nearest data container component for a given component and then calculate the key prefixes and do some `startsWith` checks that all look very fragile.